### PR TITLE
[core] Add TimeTravelUtils to resolve snapshot from scan params

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose
Add TimeTravelUtils to resolve snapshot from scan params

Linked issue: close #xxx

### CODE REVIEW

**Comment by LinMingQiang** [comment]
Expansion ：We can support more scan startup mode, like `timestamp`.

**Comment by xuzifu666** [comment]
> Expansion ：We can support more scan startup mode, like `timestamp`.
> 
> Add parameters ： `consumer.scan.startup.mode` = `snapshot/timestamp`

Thanks @LinMingQiang ,Good reminder, had added timestamp config, but mode not add due to user not need add two config if they use snapshot or timestamp.

**Comment by JingsongLi** [comment]
Maybe you want `/*+ OPTIONS('consumer-id'='myid', 'scan.snapshot-id'='3', 'consumer.ignore-progress') */`?

**Comment by xuzifu666** [comment]
> Maybe you want `/*+ OPTIONS('consumer-id'='myid', 'scan.snapshot-id'='3', 'consumer.ignore-progress' = 'true') */`?

Yes, it is a direct way, I had test for it, can query right snapshot and reset consumer. Thanks @JingsongLi

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/** Tests for {@link TimeTravelUtilsTest}. */
+public class TimeTravelUtilsTest extends ScannerTestBase {
```
Comment (LinMingQiang) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:35]: @link TimeTravelUtil

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        if (scanHandleKey.size() > 1) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "%s %s %s and %s can contains only one",
+                            CoreOptions.SCAN_SNAPSHOT_ID.key(),
+                            CoreOptions.SCAN_TAG_NAME.key(),
+                            CoreOptions.SCAN_WATERMARK.key(),
+                            CoreOptions.SCAN_TIMESTAMP_MILLIS.key()));
+        }
```
Comment (LinMingQiang) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:55]: Use `checkArgument` will be better.

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        Snapshot snapshot = null;
+        if (key.equals(CoreOptions.SCAN_SNAPSHOT_ID.key())) {
+            snapshot = handleSnapshotId(snapshotManager, options);
+        } else if (key.equals(CoreOptions.SCAN_WATERMARK.key())) {
+            snapshot = handleWatermark(snapshotManager, options);
+        } else if (key.equals(CoreOptions.SCAN_TIMESTAMP_MILLIS.key())) {
+            snapshot = handleTimestamp(snapshotManager, options);
+        } else if (key.equals(CoreOptions.SCAN_TAG_NAME.key())) {
+            snapshot = handleTagName(snapshotManager, options);
+        }
```
Comment (LinMingQiang) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:67]: switch()?

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public static List<String> getScanKeys() {
+        if (scanKeys == null) {
+            scanKeys = new ArrayList<>();
+            scanKeys.add(CoreOptions.SCAN_SNAPSHOT_ID.key());
+            scanKeys.add(CoreOptions.SCAN_TAG_NAME.key());
+            scanKeys.add(CoreOptions.SCAN_WATERMARK.key());
+            scanKeys.add(CoreOptions.SCAN_TIMESTAMP_MILLIS.key());
+        }
```
Comment (LinMingQiang) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:100]: Use String[]{} is better.

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        
+        if (scanHandleKey.size() == 0) {
+            return null;
+        }
+        if (scanHandleKey.size() > 1) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "%s %s %s and %s can contains only one",
```
Comment (LinMingQiang) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:50]: Only one of the following parameters may be set  : [%s, %s, %s]

Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java]: Thanks @LinMingQiang had addressed

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    public static String[] scanKeys = {
+        CoreOptions.SCAN_SNAPSHOT_ID.key(),
+        CoreOptions.SCAN_TAG_NAME.key(),
+        CoreOptions.SCAN_WATERMARK.key(),
+        CoreOptions.SCAN_TIMESTAMP_MILLIS.key()
+    };
+
+    public static Snapshot resolveSnapshotFromOption(
+            CoreOptions options, SnapshotManager snapshotManager) {
+        List<String> scanHandleKey = new ArrayList<>();
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:42]: List<String> scanHandleKey = new ArrayList<>(1);

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.paimon.utils.SnapshotManager;
+import org.apache.paimon.utils.TagManager;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Contains time travel functions. */
+public class TimeTravelUtil {
+
+    public static String[] scanKeys = {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:33]: private static final String[] SCAN_KEYS = {

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        CoreOptions.SCAN_SNAPSHOT_ID.key(),
+        CoreOptions.SCAN_TAG_NAME.key(),
+        CoreOptions.SCAN_WATERMARK.key(),
+        CoreOptions.SCAN_TIMESTAMP_MILLIS.key()
+    };
+
+    public static Snapshot resolveSnapshotFromOption(
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:40]: resolveSnapshotFromOptions

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            CoreOptions options, SnapshotManager snapshotManager) {
+        List<String> scanHandleKey = new ArrayList<>();
+        for (String key : scanKeys) {
+            if (options.toConfiguration().containsKey(key)) {
+                scanHandleKey.add(key);
+            }
+        }
+
+        if (scanHandleKey.size() == 0) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:49]: Add a warn log?

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            snapshot = handleWatermark(snapshotManager, options);
+        } else if (key.equals(CoreOptions.SCAN_TIMESTAMP_MILLIS.key())) {
+            snapshot = handleTimestamp(snapshotManager, options);
+        } else if (key.equals(CoreOptions.SCAN_TAG_NAME.key())) {
+            snapshot = handleTagName(snapshotManager, options);
+        }
+        return snapshot;
+    }
+
+    private static Snapshot handleSnapshotId(SnapshotManager snapshotManager, CoreOptions options) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:76]: resolveSnapshotBySnapshotId

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        Long snapshotId = options.scanSnapshotId();
+        if (snapshotId != null && snapshotManager.snapshotExists(snapshotId)) {
+            return snapshotManager.snapshot(snapshotId);
+        }
+        return null;
+    }
+
+    private static Snapshot handleTimestamp(SnapshotManager snapshotManager, CoreOptions options) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:84]: resolveSnapshotByTimestamp

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        Long timestamp = options.scanTimestampMills();
+        return snapshotManager.earlierOrEqualTimeMills(timestamp);
+    }
+
+    private static Snapshot handleWatermark(SnapshotManager snapshotManager, CoreOptions options) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:89]: resolveSnapshotByWatermark

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        Long watermark = options.scanWatermark();
+        return snapshotManager.laterOrEqualWatermark(watermark);
+    }
+
+    private static Snapshot handleTagName(SnapshotManager snapshotManager, CoreOptions options) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:94]: resolveSnapshotByTagName

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+                scanHandleKey.size() == 1,
+                String.format(
+                        "Only one of the following parameters may be set : [%s, %s, %s, %s]",
+                        CoreOptions.SCAN_SNAPSHOT_ID.key(),
+                        CoreOptions.SCAN_TAG_NAME.key(),
+                        CoreOptions.SCAN_WATERMARK.key(),
+                        CoreOptions.SCAN_TIMESTAMP_MILLIS.key()));
+
+        String key = scanHandleKey.get(0);
+        Snapshot snapshot = null;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:63]: May use switch...case better

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.paimon.CoreOptions;
+import org.apache.paimon.Snapshot;
+import org.apache.paimon.utils.Preconditions;
+import org.apache.paimon.utils.SnapshotManager;
+import org.apache.paimon.utils.TagManager;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Contains time travel functions. */
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:30]: The util class of resolve snapshot from scan params for time travel.

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        commit.commit(1, write.prepareCommit(true, 1));
+
+        write.write(rowData(3, 50, 500L));
+        commit.commit(2, write.prepareCommit(true, 2));
+
+        HashMap<String, String> optMap = new HashMap<>();
+        optMap.put("scan.snapshot-id", "2");
+        CoreOptions options = CoreOptions.fromMap(optMap);
+        Snapshot snapshot = TimeTravelUtil.resolveSnapshotFromOption(options, snapshotManager);
+        assertThat(snapshot.id() == 2);
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:57]: assertTrue

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        options = CoreOptions.fromMap(optMap);
+        snapshot = TimeTravelUtil.resolveSnapshotFromOption(options, snapshotManager);
+        assertThat(snapshot.id() == 1);
+
+        table.createTag("tag3", 3);
+        optMap.clear();
+        optMap.put("scan.tag-name", "tag3");
+        options = CoreOptions.fromMap(optMap);
+        snapshot = TimeTravelUtil.resolveSnapshotFromOption(options, snapshotManager);
+        assertThat(snapshot.id() == 3);
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:70]: assertTrue

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        optMap.put("scan.snapshot-id", "2");
+        CoreOptions options = CoreOptions.fromMap(optMap);
+        Snapshot snapshot = TimeTravelUtil.resolveSnapshotFromOption(options, snapshotManager);
+        assertThat(snapshot.id() == 2);
+
+        optMap.clear();
+        optMap.put("scan.timestamp-millis", ts + "");
+        options = CoreOptions.fromMap(optMap);
+        snapshot = TimeTravelUtil.resolveSnapshotFromOption(options, snapshotManager);
+        assertThat(snapshot.id() == 1);
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:63]: assertTrue

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java (hunk 16 above)
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:57]: add assert snapshot.id() is not null

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.HashMap;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/** Tests for {@link TimeTravelUtil}. */
+public class TimeTravelUtilsTest extends ScannerTestBase {
+
+    @Test
+    public void testScan() throws Exception {
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:38]: testResolveSnapshotFromOptions

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+        SnapshotManager snapshotManager = table.snapshotManager();
+        StreamTableWrite write = table.newWrite(commitUser);
+        StreamTableCommit commit = table.newCommit(commitUser);
+
+        write.write(rowData(1, 10, 100L));
+        commit.commit(0, write.prepareCommit(true, 0));
+        long ts = System.currentTimeMillis();
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:45]: inline this

File: paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+
+        write.write(rowData(2, 30, 101L));
+        commit.commit(1, write.prepareCommit(true, 1));
+
+        write.write(rowData(3, 50, 500L));
+        commit.commit(2, write.prepareCommit(true, 2));
+
+        HashMap<String, String> optMap = new HashMap<>();
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/table/source/snapshot/TimeTravelUtilsTest.java:53]: Map<String, String> optMap = new HashMap<>(4);

File: paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java (hunk 14 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/TimeTravelUtil.java:63]: Here may can not swith due to case key() method.

**Comment by xuzifu666** [comment]
@wwj6591812 had addressd, Thanks.

**Comment by wwj6591812** [comment]
+1

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**